docs/writing-your-first-rclrs-node.md

```diff
@@ -0,0 +1,214 @@
+# Writing your first `rclrs` node
+
+This tutorial is intended to point out some of the differences of the Rust client library with the other client libraries. It assumes knowledge of Rust, and is also not intended to be an introduction to ROS 2.
+
+As a semi-realistic example, let's create a node that periodically republishes the last message it received. It's limited to only one specific message type – `std_msgs/msg/String` in this example.
+
+## Create a package
+
+You can create a package in the usual way:
+
+```console
+cargo new republisher_node && cd republisher_node
+```
+
+In the `Cargo.toml` file, add a dependency on `rclrs = "*"` and `std_msgs = "*"`.
+
+Additionally, create a new `package.xml` if you want your node to be buildable with `colcon`. See the official [tutorial on creating a ROS 2 package](https://docs.ros.org/en/rolling/Tutorials/Creating-Your-First-ROS2-Package.html) for the structure of this file. Make sure to change the build type to `ament_cargo` and the list of dependencies to the two packages mentioned above, as such:
+
+```xml
+<depend>rclrs</depend>
+<depend>std_msgs</depend>
+
+<export>
+  <build_type>ament_cargo</build_type>
+</export>
+```
+
+
+## Writing the basic node structure
+
+Since Rust doesn't have inheritance, it's not possible to inherit from `Node` like you would in `rclcpp` or `rclpy`.
+
+Instead, you can store the node as a regular member. Let's add a struct that contains the node, a subscription, and a field for the last message that was received to `main.rs`:
```

Comment:
This is also consistent with modern guidance for writing C++. I don't understand why the official tutorials use inheritance for the Node object when composition works just as well.

Rust doesn't have inheritance (yet), but even once it does I think it would be advisable to avoid encouraging people to use inheritance in this case.

Reply:
Ok, I rephrased it to say "as is common practice in `rclcpp` or `rclpy`", so as to not endorse (or discourage) the inheritance approach in these languages.

Reply:
@tylerjw The reason the `rclcpp` and `rclpy` docs encourage inheritance is because this way initialization is homogeneous as it's delegated to the base `Node` class, and thus enables creating components without requiring to provide a custom `main` function.

It's highly unlikely that Rust will ever have inheritance, however as an improvement for the `rclrs` API we could provide default implementations for node traits:

https://doc.rust-lang.org/book/ch10-02-traits.html#default-implementations

Reply:
In order to create a class that has the interface of a node in rclcpp and therefore can be a component node all you have to do is provide one method: 

```
rclcpp::node_interfaces::NodeBaseInterface::SharedPtr get_node_base_interface()
  {
    return node_->get_node_base_interface();
  }
```

All the rest of the node interface is unnecessary for making a component node.

Reply:
It is fair to say that the details of how to use composition to achieve a component node are not documented and technically harder to understand for new developers than just inheriting from the Node... I do think this sort of thinking bleeds into all sorts of the ROS interfaces and that results in some sad tradeoffs that have been made in various interfaces.

When writing tutorials I try really hard to write idiomatic code for the language I'm using while still striving to make it as simple as possible. I don't think beginners should be encouraged to use inheritance in this way though even in tutorials.

Anyway, I don't know much about rclpy but I have been trying to fix PickNik and MoveIt's usage of C++ to be more inline with modern guidelines for using the language. I'll get off my soap box there. I did read somewhere that rust might get inheritance soon but now I can't find a link. If it doesn't, I'll be plenty happy with that as I really like the type trait system.

Reply:
I always learn something new in these reviews :) However, I'll resolve this thread since it's a bit of a tangent.